### ISSUE TITLE
Post Processor Glance / Openstack

### ISSUE THREAD

**Issue by imduffy15** [hashicorp/packer#2678]
Hey All,

I've gone through some of the past pull requests and issues, I've found two around the topic of uploading images to openstack/glance directly from packer:
- https://github.com/mitchellh/packer/issues/540
- https://github.com/mitchellh/packer/pull/2410

Neither of these have been merged though. I have started writing my own one outside of packer and would love to get feedback, improve it, and eventually merge it in with master. 

I have extended gophercloud to support the creation of images and the uploads of their binaries:
- https://github.com/monkeylittleinc/gophercloud/pull/1/files

Using this extended gophercloud I've built out a packer plugin to upload a qcow2 or raw file outputted by qemu up
- https://github.com/monkeylittleinc/packer-post-processor-glance/tree/iduffy-dev

I'm currently working on tests... Got any recommendations/examples on how to mock up the openstack service client? I took a brief look at other tests and they just seemed to avoid talking to external sources.

Thanks,
Ian.

**Comment by imduffy15** [comment]
Hey All,

I have a fully working demoable post processor working for OpenStack at the above links.

![packer run with openstack post processor](https://pbs.twimg.com/media/CNWmO7oWsAAKl3O.png:large)

I had a PR open with the folks over at gopherstack to get glance v2 support added, I've closed this PR in favour of another PR that has more extensive features. Can you guys show some community support over there for this feature? https://github.com/rackspace/gophercloud/pull/458

If successfully merged I'll update my post processor and do a PR with you guys. This seems to be a wanted feature:

#540 - Never really materialised, gophercloud version is very old.
#2410 - Preference for using the glance apis rather than SSH/SCP.

Thoughts? @vtolstov @tdhite @cbednarski

**Comment by zaletniy** [comment]
+ my vote for this feature in packer

**Comment by VMTrooper** [comment]
+1 for this feature.

**Comment by imduffy15** [comment]
Closing this, its easier to just use a post shell provisioner and the openstack glance python cli client.

**Comment by cbednarski** [comment]
@imduffy15 We've actually had quite a few requests for this. I took a look at the upstream PR earlier today and while it seems to have stalled I'd love to have native support for this in packer, so I'm going to re-open this.

**Comment by imduffy15** [comment]
The blocker is on gophercloud supporting glance interactions....... They have an open PR from https://github.com/rackspace/gophercloud/pull/458 which seems to got attention lately over here https://github.com/rackspace/gophercloud/pull/565

Until that gets merged we are kinda stuck.

**Comment by cbednarski** [comment]
@imduffy15 Thanks for the updated note! It looks like there are more updates as of today. I will check back on this. :)

**Comment by marema31** [comment]
Hello
For our openstack automatic image generation we wants to change the type of the generated image from "instance snapshot" to "image" for being more user friendly. Actually we don't find a way to do this easily and relies on a "glance image-update" command right after packer run.
It would be simplier to let packer do this since it already now the UUID of the generated artifact.

**Comment by imduffy15** [comment]
Sounds like a separate issue @marema31 this issue relates to using an offline QEMU builder to build and upload an image from packer.

From what I understand, your problem is around the openstack builder which uses some base image that is on openstack, brings up a VM, executes whatever and snapshots the VM. #3041 and #3038 would be more relevant for you and they also come with some suggested work arounds :)

**Comment by marema31** [comment]
Thanks imduffy15, I missed the point about the offline QEMU builder, and since I was trying to do actions in post-conf I rushed here. Thanks for the links, I'm going right away to have a look to them.

**Comment by jcejohnson** [comment]
@imduffy15 - Has there been any recent progress on this feature request? I need to do some post-processing on the Openstack images (i.e. -- qemu-img convert). My current solution is to do it after 'packer build' completes but that seems like the wrong thing to do versus having everything captured in my packer.json.

**Comment by imduffy15** [comment]
@jcejohnson no movement at all.... gave up on it, used the post processor shell local and a bunch of bash scripts and moved on with life.

**Comment by jcejohnson** [comment]
@imduffy15 packer version 0.10.1 does not seem to invoke my shell-local post-processor commands (https://github.com/mitchellh/packer/issues/3354). And I can't use a shell-local provisioner because the image isn't created in Openstack until the provisioner chain completes.
Were you able to work around these & stay within packer.json or are you doing some work after 'packer build' completes?

**Comment by imduffy15** [comment]
I built my openstack images offline using qemu, I pass them to a post provisioner script at the end that handles the qemu-img conversion and upload to openstack.
if you use the openstack builder you have no files, thus local shell provisioner does not run..... if it were me I'd fork packer, get the openstack builder to return the image id as a "file", send this into your post provision script, wget the image, do whatever conversions you need to do...

**Comment by jcejohnson** [comment]
Sounds similar. I build minimal kickstarted images with qemu/kvm & upload them to Openstack -- all with packer. We only need to do this when new vendor ISOs are available.
I then use another packer.json with the Openstack builder and chef-client provisioner to create our base images -- also with packer.
That bit works great until I need the qcow2 of the base image to feed into `qemu-img convert`.
I like your idea of extending the openstack builder to create a local artifact that identifies the image. I'll try to carve out some time to explore that & see where it takes me.
Thanks for the help!

**Comment by hco** [comment]
@jcejohnson you might want to look at https://www.packer.io/docs/post-processors/artifice.html :-)

**Comment by jcejohnson** [comment]
Thanks @hco. I created PR3905 to expose the image id in a file once it has been created by the Openstack builder. In my packer.json I then have artifice pick that up and make it available to shell-local where I can qemu-img convert or do other interesting things.

**Comment by cbednarski** [comment]
There are additional notes on this topic here: https://github.com/mitchellh/packer/issues/2409

**Comment by cbednarski** [comment]
Hey folks, to update here:

This issue was blocked on https://github.com/rackspace/gophercloud/pull/565 which has since been merged. Since this is a community-supported feature in packer we do not have plans to implement this, but if you have code floating around for this please feel free to open a PR and reference this issue.

Thanks!

**Comment by tdhite** [comment]
I have it all still. Let me see what I can do.
On Nov 16, 2016 7:26 PM, Chris Bednarski <notifications@github.com> wrote:Hey folks, to update here:

This issue was blocked on rackspace/gophercloud#565 which has since been merged. Since this is a community-supported feature in packer we do not have plans to implement this, but if you have code floating around for this please feel free to open a PR and reference this issue.

Thanks!

**Comment by rickard-von-essen** [comment]
Please feel free to open a PR.
